yapapi/script/__init__.py

```diff
@@ -0,0 +1,172 @@
+import asyncio
+from datetime import timedelta
+from typing import Any, Awaitable, Callable, Dict, Optional, List, Tuple, TYPE_CHECKING
+
+from yapapi.events import CommandExecuted
+from yapapi.script.capture import CaptureContext
+from yapapi.script.command import (
+    BatchCommand,
+    Command,
+    Deploy,
+    DownloadBytes,
+    DownloadFile,
+    DownloadJson,
+    Run,
+    SendBytes,
+    SendFile,
+    SendJson,
+    Start,
+    Terminate,
+)
+from yapapi.storage import DOWNLOAD_BYTES_LIMIT_DEFAULT
+
+if TYPE_CHECKING:
+    from yapapi.ctx import WorkContext
+
+
+class Script:
+    """Represents a series of commands to be executed on a provider node.
+
+    New commands are added to the script either through its `add` method or by calling one of the
+    convenience methods provided (for example: `run` or `send_json`).
+    Adding a new command *does not* result in it being immediately executed. Once ready, a `Script`
+    instance is meant to be yielded from a worker function (work generator pattern).
+    Commands will be run in the order in which they were added to the script.
+
+    If the `WorkContext` instance this `Script` uses has the field `_implicit_init` set to `True`,
+    the first script to be yielded is going to prepend the user's commands with `Deploy` and
+    `Start` commands.
+    """
+
+    timeout: Optional[timedelta] = None
+    """Time after which this script's execution should be forcefully interrupted."""
+
+    wait_for_results: bool = True
+    """Whether this script's execution should block until its results are available."""
+
+    def __init__(self, context: "WorkContext"):
+        self._ctx: "WorkContext" = context
+        self._commands: List[Tuple[Command, asyncio.Future]] = []
+
+    def _evaluate(self) -> List[BatchCommand]:
+        """Evaluate and serialize this script to a list of batch commands."""
+        batch: List[BatchCommand] = []
+        for cmd, _future in self._commands:
+            batch.append(cmd.evaluate(self._ctx))
+        return batch
+
+    async def _after(self):
+        """Hook which is executed after the script has been run on the provider."""
+        for cmd, _future in self._commands:
```

Comment:
I like to name such variables just `_` instead of `_future`. I think that's a popular convention (but I might be wrong).

Reply:
Resolved by c086011